ddtrace/contrib/vertica/patch.py

```diff
@@ -0,0 +1,269 @@
+    patch_class_routine = "{}.{}".format(patch_class, patch_routine)
+
+    @wrapt.patch_function_wrapper(patch_mod, patch_class_routine)
+    def wrapper(wrapped, instance, args, kwargs):
+        # TODO?: remove Pin dependence
+        pin = Pin.get_from(instance)
+
+        # TODO: inheritance problem with pins:
+        # Say we have a base class B and child class C:
+        # class B:
+        #   def my_method(self):
+        #     pass
+        # class C(B):
+        #   pass
+        #
+        # with tracing config:
+        #  'patch': {
+        #     'B': {
+        #       'routines': {
+        #         'my_method: {...}
+        #       }
+        #     },
+        #     'C': {}  # does not contain 'my_method'
+        #  }
+        # and an instance `c`:
+        # c = C()
+        # c.my_method()
+        #
+        # An instance of either will have a pin attached so that
+        # the user can override specific configuration for that
+        # instance.
+        #
+        # The pins will contain the config for the instance which
+        # they are attached:
+        #   PinB._config == { 'routines: { 'my_method: { ... } } }
+        #   PinC._config == { }
+        #
+        # The problem here is that at this point in the wrapper
+        # when c.my_method() is called we have an instance of C and
+        # a pin from C, but require the config from B.
+        #
+        # Possible solution: follow inheritance and look for config
+        # on a parent. Can we use normal python inheritance somehow?
+        if patch_routine in pin._config["routines"]:
+            conf = pin._config["routines"][patch_routine]
+        else:
+            conf = _find_routine_config(config, instance, patch_routine)
+
+        enabled = conf.get("trace_enabled", True)
+
+        span = None
+
+        try:
+            # shortcut if not enabled
+            if not enabled:
+                result = wrapped(*args, **kwargs)
+                return result
+
+            operation_name = conf["operation_name"]
+            tracer = pin.tracer
+            with tracer.trace(operation_name, service=pin.service) as span:
+                span.set_tags(pin.tags)
+                if "span_type" in conf:
+                    span.span_type = conf["span_type"]
+
+                if "span_start" in conf:
+                    conf["span_start"](instance, span, conf, *args, **kwargs)
+
+                result = wrapped(*args, **kwargs)
+                return result
+        except Exception as err:
+            if "on_error" in conf:
+                conf["on_error"](instance, err, span, conf, *args, **kwargs)
+            raise
+        finally:
+            # if an exception is raised result will not exist
+            if "result" not in locals():
+                result = None
+            if "span_end" in conf:
+                conf["span_end"](instance, result, span, conf, *args, **kwargs)
```

Comment:
`span` could be `None` here.

I think we want to hoist `if not enabled:` out before the `try/finally`

Reply:
Hmm. If we do this we lose the functionality calling a method after a routine when tracing is disabled. This is used in the Vertica integration for attaching data onto the cursor from the connection.

Reply:
Oh, right. I knew that.

Still have an issue of `span` being `None` if `not enabled`

Reply:
Which is also fine... since `cursor_after` doesn't use `span`

A comment in `cursor_after` might be good so anyone editing it doesn't expect a `span`.